### ISSUE TITLE
[Feature] add configuration for max window width

### ISSUE THREAD

**Issue by hertg** [leftwm/leftwm#475]
Add an **optional** configuration to set a max window width, which puts a size constraint on windows, so they don't necessarily fill the whole workspace.

Currently, every layout always tries to fill the entire screen estate, which is kind of the point of a tiling wm, but there are certain use-cases where you wouldn't want that.

*For example:*
Opening a new window on an ultra-wide monitor will fill the entire monitor with this new window, which might be alright for 34' monitors, but definitely isn't what you'd want on a 49' monitor. :)

I'm not sure if that would make sense on every layout, I think the `CenterMaster` layout is probably the one you'd use the most on such a wide monitor, but I can only speak for myself. This could also be a separate new layout like `CenterMasterMaxed` if the current layouts shouldn't behave like that.

See the graphical representation of my feature request below, based on the behavior of the `CenterMaster` layout.

**What currently happens**
```
+-----------------------------------------------+
|+---------------------------------------------+|
||                                             ||
||                     1                       ||  [49' monitor]
||                                             ||
|+---------------------------------------------+|
+-----------------------------------------------+
```

```
+-----------------------------------------------+
|+----------------------+----------------------+|
||                      |                      ||
||          1           |          2           ||  [49' monitor]
||                      |                      ||
|+----------------------+----------------------+|
+-----------------------------------------------+
```

```
+-----------------------------------------------+
|+-------------+----------------+--------------+|
||             |                |              ||
||      2      |        1       |       3      ||  [49' monitor]
||             |                |              ||
|+-------------+----------------+--------------+|
+-----------------------------------------------+
```

**What I'd want to happen**
```
+-----------------------------------------------+
|               +---------------+               |
|               |               |               |
|               |       1       |               |  [49' monitor]
|               |               |               |
|               +---------------+               |
+-----------------------------------------------+

                ^^^^^^^^^^^^^^^^^
                MAX_WINDOW_WIDTH
```

```
+-----------------------------------------------+
|        +--------------+--------------+        |
|        |              |              |        |
|        |       1      |       2      |        |  [49' monitor]
|        |              |              |        |
|        +--------------+--------------+        |
+-----------------------------------------------+

         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              MAX_WINDOW_WIDTH * 2
```

```
+-----------------------------------------------+
|+-------------+----------------+--------------+|
||             |                |              ||
||      2      |        1       |       3      ||  [49' monitor]
||             |                |              ||
|+-------------+----------------+--------------+|
+-----------------------------------------------+
```

I'd be happy to provide a PR with this feature, but I need some feedback of the maintainers whether this is something they even want or not.

Coming from dwm, where I created a custom CenteredMasterLayout which behaved like that, this and #474 are the main things that I currently miss on LeftWM. Other than that, LeftWM is great, and I'm looking forward to make the switch on all my computers! :)

Btw: As I'm writing this, I have two empty terminals open, just to have my browser window at an usable width :stuck_out_tongue: 
![2021-09-12_11-44-56](https://user-images.githubusercontent.com/3533648/132983209-e405398e-70b7-4653-aaa5-080d455ea451.png)

**A note about workspaces**
I'm aware of the multiple workspaces per monitor feature, but I'd prefer to use the feature in my proposal. 
AFAIK workspaces are hard-coded in the configuration with the exact width/x/y (correct me if I'm wrong). I have two issues with that:
- There is no easy way to dynamically resize the master window width. This is also not possible right now, but I provided a fix for that in #474 
- I often work on my laptop where I change places frequently (undocked, docked to normal monitor, docked to ultrawide, etc.), a statically configured workspace doesn't really work in such a dynamic use-case.

**Comment by VuiMuich** [comment]
Welcome Adventurer!
I can see that this might be an interesting and helpful feature. (and this is a very well written feature request, btw. :+1: )

Similar to my proposal in [Layout Per Tag #448](https://github.com/leftwm/leftwm/pull/448#), I would suggest to introduce this as an optional field for workspaces, when static configured and as a general optional field, when workspaces are kept dynamic.
So either would be valid:
```toml
workspaces = []
# no field for max_window_witdth
```
```toml
workspaces = []
max_window_witdth = 1920
```
```toml
#note: not sure if a screen with 3840*1440 actually exists...
[[workspaces]]
y = 0
x = 0
height = 1440
width = 1280

[[workspaces]]
y = 0
x = 1280
height = 1440
width = 2560
max_window_width = 1280
```

*side note: maybe introducing the possibility to config workspaces in either percentage or pixel (defaulting to pixel if not specifically set) would be nice, and at some point maybe some possibilities for a bit of logic for dynamic workspaces...

**Comment by hertg** [comment]
Your suggestion regarding the configuration sounds great, and to make it configurable per workspace would be pretty neat. I would definitely add this as an optional configuration, where the value defaults to `-1`  (meaning no maximum width) or similar if none is found in the config.

> This might actually be a nice second PR :wink:

Sounds great. As already pointed out, I am heavily reliant on that feature for my niche "laptop docked to 49' monitor" use-case. So I can probably start working on that in the next few days/weeks if that's okay with everyone. :)

**Comment by VuiMuich** [comment]
> Your suggestion regarding the configuration sounds great, and to make it configurable per workspace would be pretty neat. I would definitely add this as an optional configuration, where the value defaults to `-1`  (meaning no maximum width) or similar if none is found in the config.

I believe in case the option is not set in the config defaulting to `max_window_size` to `workspace.w` (if I get the width field right from the top of my head) would be the most sensible thing. 

> > This might actually be a nice second PR :wink:
> 
> 
> 
> Sounds great. As already pointed out, I am heavily reliant on that feature for my niche "laptop docked to 49' monitor" use-case. So I can probably start working on that in the next few days/weeks if that's okay with everyone. :)

I guess that's totally fine, as you have probably the most urgent use case yourself 😆

**Comment by hertg** [comment]
closed with PR #482
